seal.go

```diff
@@ -135,6 +155,148 @@ type KeyCreationParams struct {
+// to SealedKeyObject.UnsealFromTPM.
+//
+// The authorization key can also be chosen and provided by setting
+// AuthKey in the params argument.
+func SealKeyToExternalTPMStorageKey(tpmKey *tpm2.Public, key []byte, keyPath string, params *KeyCreationParams) (authKey TPMPolicyAuthKey, err error) {
+	// params is mandatory.
+	if params == nil {
+		return nil, errors.New("no KeyCreationParams provided")
+	}
+
+	// Perform some sanity checks on params.
+	if params.AuthKey != nil && params.AuthKey.Curve != elliptic.P256() {
+		return nil, errors.New("provided AuthKey must be from elliptic.P256, no other curve is supported")
+	}
+
+	if params.PCRPolicyCounterHandle != tpm2.HandleNull {
+		return nil, errors.New("PCRPolicyCounter must be tpm2.HandleNull when creating an importable sealed key")
+	}
+
+	succeeded := false
+
+	// Compute metadata.
+
+	var goAuthKey *ecdsa.PrivateKey
+	// Use the provided authorization key,
+	// otherwise create an asymmetric key for signing
+	// authorization policy updates, and authorizing dynamic
+	// authorization policy revocations.
+	if params.AuthKey != nil {
+		goAuthKey = params.AuthKey
+	} else {
+		goAuthKey, err = ecdsa.GenerateKey(elliptic.P256(), rand.Reader)
+		if err != nil {
+			return nil, xerrors.Errorf("cannot generate key for signing dynamic authorization policies: %w", err)
+		}
+	}
+	authPublicKey := createTPMPublicAreaForECDSAKey(&goAuthKey.PublicKey)
+	authKey = goAuthKey.D.Bytes()
+
+	pub := makeImportableSealedKeyTemplate()
+
+	// Compute the static policy - this never changes for the lifetime of this key file
+	staticPolicyData, authPolicy, err := computeStaticPolicy(pub.NameAlg, &staticPolicyComputeParams{key: authPublicKey})
+	if err != nil {
+		return nil, xerrors.Errorf("cannot compute static authorization policy: %w", err)
+	}
+
+	pub.AuthPolicy = authPolicy
+
+	// Create a dynamic authorization policy
+	pcrProfile := params.PCRProfile
+	if pcrProfile == nil {
+		pcrProfile = &PCRProtectionProfile{}
+	}
+	dynamicPolicyData, err := computeSealedKeyDynamicAuthPolicy(nil, currentMetadataVersion, pub.NameAlg, authPublicKey.NameAlg,
+		goAuthKey, nil, nil, pcrProfile, nil)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot compute dynamic authorization policy: %w", err)
+	}
+
+	// Clean up files on failure.
+	defer func() {
+		if succeeded {
+			return
+		}
+		os.Remove(keyPath)
+	}()
+
+	// Seal key
+
+	// Create the destination file
+	f, err := os.OpenFile(keyPath, os.O_WRONLY|os.O_CREATE|os.O_EXCL, 0600)
+	if err != nil {
+		return nil, xerrors.Errorf("cannot create key data file: %w", err)
+	}
+	defer f.Close()
+
+	// Create the sensitive data
+	sealedData, err := mu.MarshalToBytes(sealedData{Key: key, AuthPrivateKey: authKey})
+	if err != nil {
+		panic(fmt.Sprintf("cannot marshal sensitive data: %v", err))
+	}
+	sensitive := tpm2.Sensitive{
+		Type:      pub.Type,
+		AuthValue: make(tpm2.Auth, pub.NameAlg.Size()),
```

Comment:
should this have a comment that no authorization is set?

Reply:
I can actually omit setting this field here - I forgot that it gets padded automatically in go-tpm2 instead: https://github.com/canonical/go-tpm2/blob/master/utils.go#L461